### PR TITLE
40384: Fix Chart Filtering

### PR DESCRIPTION
**Pull request by labkey-nicka**
#### Rationale
This PR addresses issues [37153](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=37153) and [40384](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=40384) by clearly delineating when and how filters are applied to charts from the different sources.

#### Changes
* Refactor how the set of filters is constructed that fulfills the query configuration filters backing charts.
* Switch to applying both immutable and mutable (aka "removable") filters to the associated QueryWebPart (QWP).
* Outline (as a comment) the supported filtering sources.

### CODE REVIEW

File: visualization/resources/web/vis/chartWizard/genericChartPanel.js [hunk 1]
```diff
@@ -663,29 +664,38 @@ Ext4.define('LABKEY.ext4.GenericChartPanel', {
...
+            // If the region does not exist, then apply it's filters from the URL
+            // and allow them to be mutated on the QWP.
+            removableFilters = removableFilters.concat(this.getDataRegionFilters());
+        }
 
-        var userSort = LABKEY.Filter.getSortFromUrl(filterUrl, this.dataRegionName);
+        var userSort = LABKEY.Filter.getSortFromUrl(this.getFilterURL(), this.dataRegionName);
 
-        this.currentFilterStr = this.createFilterString(userFilters);
+        this.currentFilterStr = this.createFilterString(removableFilters);
```
Comment (cnathe) [inline visualization/resources/web/vis/chartWizard/genericChartPanel.js]: since `this.currentFilterStr` is used to determine if the chart needs to be re-rendered when the user switches between the "View Data" and "View Chart" button options, I think this needs to use the `filters` instead of `removableFilters`:
```suggestion
        this.currentFilterStr = this.createFilterString(filters);
```

File: visualization/resources/web/vis/chartWizard/genericChartPanel.js [hunk 2]
```diff
@@ -778,20 +814,13 @@ Ext4.define('LABKEY.ext4.GenericChartPanel', {
             config.scope = this;
         }
 
-        var filters;
+        // Filter scenarios (Issue 37153, Issue 40384)
+        // 1. Filters defined explicitly on chart configuration (this.userFilters)
+        // 2. Filters defined on associated QWP (panelDataRegionName)
+        // 3. Filters defined on associated Data Region (dataRegionName)
+        // 4. Filters defined on URL for associated Data Region (overlap with #3)
```
Comment (cnathe) [inline visualization/resources/web/vis/chartWizard/genericChartPanel.js:838]: Thanks for writing this up and including it in a comment. After this issue is resolved and PR merged, we should probably pass this over to the test team as a TODO for adding a regression test for the various combinations of these filter scenarios.

**Review by cnathe — APPROVED** [review]
Overall looks good. Manual testing done locally with charts that have saved filters as part of their config and charts without. Then I tested applying this run URL filter on top of both and from there adding/removing "View Data" QWP filters. All looks good.

The one comment about the "should the chart re-render" check was all I found, but approving as no need to re-review.

File: visualization/resources/web/vis/chartWizard/genericChartPanel.js (hunk 2 above)
Comment (labkey-nicka) [reply visualization/resources/web/vis/chartWizard/genericChartPanel.js:838]: I've pointed Chris to these and we've gone over these scenarios. He is going to take on updating the Visualization tests that need to be updated due to this and we'll work through ensuring coverage for these scenarios.

File: visualization/resources/web/vis/chartWizard/genericChartPanel.js (hunk 1 above)
Comment (labkey-nicka) [reply visualization/resources/web/vis/chartWizard/genericChartPanel.js]: It ended up being even more subtle than this. I've updated this to include all filters since the user could modify any of these filters (e.g. either on the Data Region directly or on the QWP directly). See 1ba0ae4.

File: visualization/resources/web/vis/chartWizard/genericChartPanel.js (hunk 2 above)
Comment (labkey-nicka) [reply visualization/resources/web/vis/chartWizard/genericChartPanel.js:838]: I ended up triaging the tests and they had found edge cases with the `filterUrl` parameter. I've since patched with 3d4d6ce and the tests are looking good.
